packages/alfa-aria/src/dom/predicate/is-programmatically-hidden.ts

```diff
@@ -34,39 +34,32 @@ export function isProgrammaticallyHidden(
   );
 }
 
-const cache = Cache.empty<Device, Cache<Context, Cache<Node, boolean>>>();
-
-function hasHiddenAncestors(
+function _hasHiddenAncestors(
   device: Device,
-  context: Context = Context.empty(),
+  context: Context,
 ): Predicate<Node> {
-  return (node) =>
-    cache
-      .get(device, Cache.empty)
-      .get(context, Cache.empty)
-      .get(node, () =>
-        test(
-          or(
-            // Either it is a programmatically hidden element
-            and(
-              isElement,
-              or(
-                hasComputedStyle(
-                  "display",
-                  ({ values: [outside] }) => outside.value === "none",
-                  device,
-                  context,
-                ),
-                hasAttribute("aria-hidden", equals("true")),
-              ),
-            ),
-            // Or its parent is programmatically hidden
-            (node: Node) =>
-              node
-                .parent(Node.fullTree)
-                .some(hasHiddenAncestors(device, context)),
-          ),
-          node,
+  return or(
+    // Either it is a programmatically hidden element
+    and(
+      isElement,
+      or(
+        hasComputedStyle(
+          "display",
+          ({ values: [outside] }) => outside.value === "none",
+          device,
+          context,
         ),
-      );
+        hasAttribute("aria-hidden", equals("true")),
+      ),
+    ),
+    // Or its parent is programmatically hidden
+    (node: Node) =>
+      node.parent(Node.fullTree).some(_hasHiddenAncestors(device, context)),
+  );
 }
+
+const hasHiddenAncestors = Cache.memoize<
+  unknown,
+  [Device, Context],
+  Predicate<Node>
+>(_hasHiddenAncestors);
```

Comment:
I'm not sure why TS needs some skaffolding here, but not in the simple cases in the tests 🤷